File: MainPart/src/day02/array.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

typedef struct arr
{
    int *content;
    int size;
    // 已用容量
    int capacity;
    // 理论容量
} arr;

arr* create_arr(int capacity) {
    arr *array = (arr*) malloc(sizeof(arr));  // 分配结构体本身
    array->size = 0;
    array->capacity = capacity;
    array->content = (int*) calloc(capacity, sizeof(int));  // 分配元素空间
    // 这里要注意，不可以前面都写array，结果后面返回对array取址
    // 因为那样的话array就是临时变量，等函数执行完就销毁了
    return array;
}

arr* resize(arr* array, int new_size) {
    arr* new_array = create_arr(new_size);

    for (int i = 0; i < array->size; i++) new_array->content[i] = array->content[i];
    new_array -> size = array -> size;

    free(array->content);
    free(array);

    return new_array;
}


bool is_element_index(arr* array, int index) {
    return (0 <= index)  && (index < array->size);
}

bool is_position_index(arr* array, int index) {
    return (0 <= index)  && (index <= array->size);
}

int get_size(arr* array) {
    return array->size;
}

void show_array(arr* array) {
    printf("****************************\n");
    printf("Size: %d | Capacity: %d\n", array -> size, array -> capacity);
    for (int i = 0; i < array-> size; i++) printf("number%2d -%2d ;\n", i+1, array->content[i]);
    return ;
}


arr* append(arr* array, int element) {
    if (array->size == array->capacity) array = resize(array, 2 * array->capacity);
    array -> content[array -> size] = element;
    array -> size ++;
    return array;
}

arr* add(arr* array, int index, int element){
    if (!is_position_index(array, index)) return array;

    if (array->size == array->capacity) array = resize(array, 2 * array->capacity);

    if (index < array->size) {
        for (int j = array -> size; j > index; j-- ) {
            array -> content [j] = array -> content [j-1];
        } 
        array -> content[index] = element;
        array->size++;
    } else if (index == array->size) {
        array = append(array, element);
    } else {
        return array;
    }


    return array;
}

arr* pop(arr* array) {
    array->size --;
    if (array->size<(array->capacity/4)) array = resize(array, array->capacity/2);
    array->content[array->size-1] = 0;
    return array;
}

arr* remove_index(arr* array, int index) {
    if (!is_position_index(array, index)) return array;
    if (array->size<(array->capacity/4)) array = resize(array, array->capacity/2);

    if (index < array->size) {
        for (int j = index; j < array -> size-1; j++ ) {
            array -> content [j] = array -> content [j+1];
        } 
        array -> content[array->size] = 0;
        array->size--;
    } else if (index == array->size) {
        array = pop(array);
    } else {
        return array;
    }

    return array;
}

int main() {
    arr *array = create_arr(5);

    for (int i = 0; i < 10; i++) {
        array = append(array, i);
    }

    show_array(array);
    
    array = add(array, 2, 666);
    array = add(array, 11, 666);
    show_array(array);

    array = pop(array);
    show_array(array);

    array = remove_index(array, 2);
    show_array(array);
    return 0;
}

File: MainPart/src/day02/array.py
class MyArray:
    def __init__(self, init_capacity: int = 1):
        self.data = [None for i in range(init_capacity)]
        self.size = 0

    def append(self, e):
        capacity = len(self.data)
        if self.size == capacity:
            self._resize(2 * capacity)
        self.data[self.size] = e
        self.size += 1
        return 

    def add(self, index, e):
        # 检查索引越界
        self._check_position_index(index)

        cap = len(self.data)
        # 看 data 数组容量够不够
        if self.size == cap:
            self._resize(2 * cap)

        # 搬移数据 data[index..] -> data[index+1..]
        # 给新元素腾出位置
        for i in range(self.size-1, index-1, -1):
            self.data[i+1] = self.data[i]
        
        # 插入新元素
        self.data[index] = e

        self.size += 1

    def add_first(self, e):
        self.add(0, e)

    # 删
    def remove_last(self):
        if self.size == 0:
            raise Exception("NoSuchElementException")
        cap = len(self.data)
        # 可以缩容，节约空间
        if self.size == cap // 4:
            self._resize(cap // 2)

        deleted_val = self.data[self.size - 1]
        # 删除最后一个元素
        self.data[self.size - 1] = None
        self.size -= 1

        return deleted_val

    def remove(self, index):
        # 检查索引越界
        self._check_element_index(index)

        cap = len(self.data)
        # 可以缩容，节约空间
        if self.size == cap // 4:
            self._resize(cap // 2)

        deleted_val = self.data[index]

        # 搬移数据 data[index+1..] -> data[index..]
        for i in range(index + 1, self.size):
            self.data[i - 1] = self.data[i]

        self.data[self.size - 1] = None
        self.size -= 1

        return deleted_val

    def remove_first(self):
        return self.remove(0)

    # 查
    def get(self, index):
        # 检查索引越界
        self._check_element_index(index)

        return self.data[index]

    # 改
    def set(self, index, element):
        # 检查索引越界
        self._check_element_index(index)
        # 修改数据
        old_val = self.data[index]
        self.data[index] = element
        return old_val

    # 工具方法
    def get_size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    # 将 data 的容量改为 newCap
    def _resize(self, new_cap):
        temp = [None] * new_cap
        for i in range(self.size):
            temp[i] = self.data[i]
        self.data = temp

    def _is_element_index(self, index):
        return 0 <= index < self.size

    def _is_position_index(self, index):
        return 0 <= index <= self.size

    def _check_element_index(self, index):
        if not self._is_element_index(index):
            raise IndexError(f"Index: {index}, Size: {self.size}")

    def _check_position_index(self, index):
        if not self._is_position_index(index):
            raise IndexError(f"Index: {index}, Size: {self.size}")

    def display(self):
        print(f"size = {self.size}, cap = {len(self.data)}")
        print(self.data)


# Usage example
if __name__ == "__main__":
    arr = MyArray(init_capacity=3)

    # 添加 5 个元素
    for i in range(1, 6):
        arr.append(i)

    arr.remove(3)
    arr.add(1, 9)
    arr.add_first(100)
    val = arr.remove_last()

    # 100 1 9 2 3
    for i in range(arr.get_size()):
        print(arr.get(i))